File: PCPextractor.m
file = 'c2.wav' ;
[y,Fs,bits] = wavread(file); 

N = length(y); 
t = (1/Fs)*(1:N); 
y_fft = abs(fft(y));            
y_fft = y_fft(1:N/2)  ;    
PCP = zeros(12, 1) ;
fref = 130.8 ;
M = zeros(N/2 , 1) ;
for l = 1:N/2,
    if l == 1,
      M(l) = -1 ;
    else,
      M(l) = mod(round(12*log2((Fs*((l-1)/N))/fref)), 12) ;
    endif
endfor 
for p = 1:12,
  PCP(p) = ((y_fft.^2)')*kroneckerDel(M, (p-1)*ones(N/2, 1)) ;
endfor

PCP = PCP./sum(PCP) 

File: jim2012Chords/Guitar_Only/PCPextractor.py
import numpy, scipy.io.wavfile
from matplotlib import pyplot
print("Enter file path")
path=input()
fs, y = scipy.io.wavfile.read(path)
n=numpy.size(y)
k=int(n/2)
y=numpy.transpose(numpy.square(abs(numpy.fft.fft(y))[:k]))
pcp=numpy.zeros(k)
fref=130.8
M=numpy.zeros(k)
M[0]=-1
for l in range(1,k):
	M[l] = round(12*numpy.log2((fs/fref)*(l/n)))%12
pcp=numpy.zeros(12)
for i in range(12):
	pcp[i]=numpy.dot(y,(M==(i*numpy.ones(k))))
pcp/=numpy.sum(pcp)
with pyplot.xkcd():
	pyplot.plot(pcp)
pyplot.show()